test/bin/ci_phase_iso_build.sh

```diff
@@ -21,49 +20,111 @@ exec &> >(tee >(awk '{ print strftime("%Y-%m-%d %H:%M:%S"), $0; fflush() }' >"${
+            $(dry_run) bash -x ./bin/build_images.sh ${build_opts} -g ./image-blueprints/group3
+        fi
+    else
+        # Fall back to full build when not running in CI
+        $(dry_run) bash -x ./bin/build_images.sh ${build_opts}
+    fi
+}
+
 # Clean the dnf cache to avoid corruption
-sudo dnf clean all
+$(dry_run) sudo dnf clean all
 
 # Show what other dnf commands have been run to try to debug why we
 # sometimes see cache collisons.
-sudo dnf history --reverse
+$(dry_run) sudo dnf history --reverse
 
 cd "${ROOTDIR}"
 
 # Get firewalld and repos in place. Use scripts to get the right repos
 # for each branch.
-bash -x ./scripts/devenv-builder/configure-vm.sh --no-build --force-firewall "${PULL_SECRET}"
-bash -x ./scripts/image-builder/configure.sh
+$(dry_run) bash -x ./scripts/devenv-builder/configure-vm.sh --no-build --force-firewall "${PULL_SECRET}"
+$(dry_run) bash -x ./scripts/image-builder/configure.sh
 
 cd "${ROOTDIR}/test/"
 
 # Re-build from source.
-bash -x ./bin/build_rpms.sh
+$(dry_run) bash -x ./bin/build_rpms.sh
 
 # Set up for scenario tests
-bash -x ./bin/create_local_repo.sh
+$(dry_run) bash -x ./bin/create_local_repo.sh
 
 # Start the web server to host the ostree commit repository for parent images
-bash -x ./bin/start_webserver.sh
+$(dry_run) bash -x ./bin/start_webserver.sh
 
-# Build all of the images
+# Figure out an optimal number of osbuild workers
 CPU_CORES="$(grep -c ^processor /proc/cpuinfo)"
 MAX_WORKERS=$(find "${ROOTDIR}/test/image-blueprints" -name \*.toml | wc -l)
 CUR_WORKERS="$( [ "${CPU_CORES}" -lt  $(( MAX_WORKERS * 2 )) ] && echo $(( CPU_CORES / 2 )) || echo "${MAX_WORKERS}" )"
 
-bash -x ./bin/start_osbuild_workers.sh "${CUR_WORKERS}"
+$(dry_run) bash -x ./bin/start_osbuild_workers.sh "${CUR_WORKERS}"
 
-# Image build can be optimized in CI based on the job type
-if [ -v CI_JOB_NAME ] ; then
-    bash -x ./bin/build_images.sh -g ./image-blueprints/group1
-    bash -x ./bin/build_images.sh -g ./image-blueprints/group2
-    # Group 3 only contains images used in periodic CI jobs
-    if [[ "${CI_JOB_NAME}" =~ .*periodic.* ]]; then
-        bash -x ./bin/build_images.sh -g ./image-blueprints/group3
+# Check if cache can be used for builds
+# This will fail when AWS S3 connection is not configured, or there is no cache bucket
+HAS_BUILD_CACHE=false
+if ./bin/manage_build_cache.sh getlast -b "${SCENARIO_BUILD_BRANCH}" -t "${SCENARIO_BUILD_TAG}" &>/dev/null ; then
+    HAS_BUILD_CACHE=true
+fi
+
+# Check the build mode: "use cache" (default) or "update cache"
+if [ $# -gt 0 ] && [ "$1" = "-update_cache" ] ; then
+    if ! ${HAS_BUILD_CACHE} ; then
+        echo "ERROR: Access to the build cache is not available"
+        exit 1
     fi
+    build_cache_update
 else
-    # Fall back to full build when not running in CI
-    bash -x ./bin/build_images.sh
+    if ! ${HAS_BUILD_CACHE} ; then
+        echo "Build cache is not available, rebuilding all the artifacts"
+    fi
+    build_with_cache
```

Comment:
Nit: slightly confusing naming: `not $HAS_BUILD_CACHE` and then `build_with_cache`. Maybe that `echo` could be an `else` to the `if ${HAS_BUILD_CACHE}` inside `build_with_cache`?